File: itx-rs/itx-backend/src/feature/post/use_case/list_posts.rs
use crate::error::BackendError;
use itx_contract::repo::post::PostRepo;
use serde::Serialize;
use std::sync::Arc;
use uuid::Uuid;

pub struct ExecuteParams {
    pub user_id: Uuid,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExecuteOutput {
    pub items: Vec<Item>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Item {}

pub struct ListPostsUseCase {
    post_repo: Arc<dyn PostRepo>,
}

impl ListPostsUseCase {
    pub fn new(post_repo: Arc<dyn PostRepo>) -> Self {
        Self { post_repo }
    }

    pub async fn execute(&self, params: ExecuteParams) -> Result<ExecuteOutput, BackendError> {
        Ok(ExecuteOutput { items: vec![] })
    }
}

File: itx-ts/src/backend/feature/post/use_case/list_posts.ts
import { PostDto, PostRepo } from "../../../../types.js";
import { toPostDto } from "../../../../util.js";

export type ExecuteParams = {
  userId: string;
  limit: number;
  offset: number;
};

export type ExecuteOutput = {
  items: PostDto[];
};

export class ListPostsUseCase {
  constructor(private readonly postRepo: PostRepo) {}

  async execute(params: ExecuteParams): Promise<ExecuteOutput> {
    const posts = await this.postRepo.list({
      authorId: params.userId,
      limit: params.limit,
      offset: params.offset,
    });
    return { items: posts.map(toPostDto) };
  }
}
